src/tools/cephfs_mirror/SnapReplicator.cc

```diff
@@ -0,0 +1,434 @@
+          is_error = true;
+          break;
+        }
+      }
+    }
+    if (ceph_closedir(m_src_mnt, new_dirp) < 0) {
+      std::clog << __func__ << ":" << __LINE__ << ": closedir failed for " << ("." + dir_name) << std::endl;
+    }
+    new_dirp = NULL;
+  }
+  if (!is_error) {
+    int rv = 0;
+    std::string dst_snap = m_src_dir + "/"s + m_dst_snap_dir + "/"s + m_new_snap;
+    if ((rv = ceph_mkdir(m_dst_mnt, dst_snap.c_str(), 0755)) < 0 and rv != -EEXIST) {
+      std::clog << __func__ << ":" << __LINE__ << ": failed to create dst snapshot dir " << dst_snap << std::endl;
+      my_errno = -rv;
+      return -1;
+    }
+    return 0;
+  }
+  return -1;
+}
+
+struct timespec SnapReplicator::get_rctime(const std::string& snap_name, const std::string& dir)
+{
+  struct timespec ts{0, 0};
+
+  if (snap_name == ""s)
+    return ts;
+
+  char buf[64] = {0,};
+  if (ceph_lgetxattr(m_src_mnt, dir.c_str(), "ceph.dir.rctime", buf, sizeof(buf)) < 0) {
+    std::clog << __func__ << ":" << __LINE__ << ": lgetxattr failed for: " << dir << "\n";
+    return ts;
+  }
+
+  const char *dot = strchr(buf, '.');
+  ceph_assert(dot);
+  std::string sec = std::string(buf, dot - buf);
+  std::string nsec = std::string(dot + 1);
+  ts.tv_sec = std::stol(sec);
+  ts.tv_nsec = std::stol(nsec);
+  return ts;
+}
+
+bool SnapReplicator::is_system_dir(const char *dir) const
+{
+  if ((dir[0] == '.' && dir[1] == '\0') ||
+      (dir[0] == '.' && dir[1] == '.' && dir[2] == '\0'))
+    return true;
+
+  return false;
+}
+
+int SnapReplicator::copy_remaining(int read_fd, int write_fd, off_t read, int len)
+{
+  bool is_error = false;
+  int rv = 0;
+
+  while (read < len) {
+    if ((rv = ceph_read(m_src_mnt, read_fd, m_readbuf_new, TXN_BLOCK_SIZE, read)) < 0) {
+      std::clog << __func__ << ":" << __LINE__ << ": source file read with ceph_read(" << read_fd << ", " << TXN_BLOCK_SIZE << ", " << read << ") failed\n";
+      my_errno = -rv;
+      is_error = true;
+      break;
+    }
+    int read_bytes = rv;
+    if ((rv = ceph_write(m_dst_mnt, write_fd, m_readbuf_new, rv, read)) < 0) {
+      std::clog << __func__ << ":" << __LINE__ << ": destination file write with ceph_write(" << write_fd << ", " << read_bytes << ", " << read << ") failed\n";
+      my_errno = -rv;
+      is_error = true;
+      break;
+    }
+    read += rv;
+  }
+  if (is_error)
+    return -1;
+  return 0;
+}
+
+int SnapReplicator::copy_blocks(const std::string& en, file_type_t ftype,
+                                const struct ceph_statx& old_stx,
+                                const struct ceph_statx& new_stx)
+{
+  bool is_error = false;
+
+  if (ftype == file_type_t::OLD_FILE) {
+    off_t old_size = old_stx.stx_size;
+
+    if (old_size == 0) {
+      std::clog << __func__ << ":" << __LINE__ << ": old file size is zero; copying all new file blocks\n";
+      goto copy_all_blocks;
+    }
+
+    std::string old_root = m_src_dir + "/"s + m_src_snap_dir + "/"s + m_old_snap;
```

Comment:
Would it be better to have a helper function for this? Seems to be used in multiple places...